### PR TITLE
Self-service errors: generate error deprecation docs [KVL-1187]

### PR DESCRIPTION
**Pull request by fabiotudone-da**
A new "Deprecated" section is rendered for errors that have `@deprecated` with a message.

[Here's the preview](https://htmlpreview.github.io/?https://raw.githubusercontent.com/fabiotudone-da/kvl-1187-html-preview/main/html/error-codes/self-service/index.html).

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by fabiotudone-da** [comment]
The build fails due to the Scala Reflection API being different in Scala 2.12 and 2.13, looking into that.

File: ledger/error/src/main/scala/com/daml/error/generator/ErrorCodeDocumentationGenerator.scala [hunk 1]
```diff
@@ -104,38 +106,52 @@ class ErrorCodeDocumentationGenerator(prefixes: Array[String] = Array("com.daml"
...
+          sys.error(
+            s"Unexpected annotation detected (${annotations.map(annotationTypeName)} but the only supported ones are $acceptedTypeNames)."
+          )
     }
 
-    ErrorDocumentationAnnotations(doc.explanation, doc.resolution)
+    ErrorDocumentationAnnotations(doc.deprecation, doc.explanation, doc.resolution)
   }
 
+  private def parseDeprecatedAnnotationValue(tree: ru.Tree): String =
```
Comment (hubert-da) [inline ledger/error/src/main/scala/com/daml/error/generator/ErrorCodeDocumentationGenerator.scala:142]: wow

**Review by hubert-da — APPROVED** [review]
LGTM, however, you may want to wait for other reviewers

**Review by tudor-da — APPROVED** [review]

File: ledger/error/src/main/scala/com/daml/error/generator/ErrorCodeDocumentationGenerator.scala (hunk 1 above)
Comment (fabiotudone-da) [reply ledger/error/src/main/scala/com/daml/error/generator/ErrorCodeDocumentationGenerator.scala:181]: Yeah, Java reflection would be easier but it doesn't seem to help with Scala's `@deprecated` (the properties are not accessible).

**Review by cocreature — APPROVED** [review]
Scala version guards look fine, we build & push the docs from a 2.13 build so this shouldn’t break anything.

**Review by cocreature — APPROVED** [review]
Scala version guards look fine, we build & push the docs from a 2.13 build so this shouldn’t break anything.

**Comment by fabiotudone-da** [comment]
There is still a remaining problem in that we build the generator also as part of the `/ledger/error` library (and libraries are cross-compiled).

**Comment by cocreature** [comment]
> There is still a remaining problem in that we build the generator also as part of the /ledger/error library (and libraries are cross-compiled).

**Comment by fabiotudone-da** [comment]
> If it does need to be in 1.18 happy to have a chat on possible solutions. We can exclude some stuff from being cross-compiled & published in pricniple.

Yes it should be in 1.18, I'm splitting off the generator and that should solve the issue; if not, I'll leave it to you @tudor-da :)

**Pull request merged**